packages/frontend/src/notebook/notebook_editor.tsx

```diff
@@ -249,60 +247,74 @@ export function NotebookEditor<T>(props: {
                 </div>
             </Show>
             <ul class="notebook-cells">
-                <For each={props.notebook.cells}>
-                    {(cell, i) => {
+                <For each={props.notebook.cellOrder}>
+                    {(cellId, i) => {
+                        const cell = props.notebook.cellContents[cellId];
+                        if (!cell) {
+                            throw `Failed to find contents for cell ${cellId}`;
+                        }
+
                         const isActive = () => activeCell() === i();
                         const cellActions: CellActions = {
                             activateAbove() {
                                 i() > 0 && setActiveCell(i() - 1);
                             },
                             activateBelow() {
-                                const n = props.notebook.cells.length;
+                                const n = props.notebook.cellOrder.length;
                                 i() < n - 1 && setActiveCell(i() + 1);
                             },
                             createAbove() {
                                 props.changeNotebook((nb) => {
-                                    nb.cells.splice(i(), 0, newStemCell());
+                                    const newCell = newStemCell();
+                                    nb.cellOrder.splice(i(), 0, newCell.id);
+                                    nb.cellContents[newCell.id] = newCell;
```

Comment:
Though I know it's sort of trivial, this pair of lines is appearing so frequently that it might be worthwhile to have a helper function `insertCell(notebook, i, cell)` in `src/notebook/types.ts`.